File: src/test/java/com/hileco/cortex/primitives/test/Variation.java
package com.hileco.cortex.primitives.test;

import java.util.Random;
import java.util.function.Consumer;

public class Variation {
    private Random random;

    public Variation(long seed) {
        this.random = new Random(seed);
    }

    public void seed(long seed) {
        this.random = new Random(seed);
    }

    public void maybe(Runnable runnable) {
        if (random.nextBoolean()) {
            runnable.run();
        }
    }

    public static void fuzzed(long times, Consumer<Variation> fuzzee) {
        for (var i = 0L; i < times; i++) {
            try {
                var variation = new Variation(i);
                fuzzee.accept(variation);
            } catch (Throwable t) {
                throw new IllegalStateException(String.format("Exception using variation seed %d of %d", i, times), t);
            }
        }
    }
}

File: cortex-collections/src/test/kotlin/com/hileco/cortex/collections/test/Variation.kt
package com.hileco.cortex.collections.test

import java.util.*

class Variation(seed: Long) {
    private var random: Random? = Random(seed)

    fun seed(seed: Long) {
        this.random = Random(seed)
    }

    fun maybe(runnable: () -> Unit) {
        if (random!!.nextBoolean()) {
            runnable()
        }
    }

    companion object {
        fun fuzzed(times: Long, fuzzee: (Variation) -> Unit) {
            for (i in 0L until times) {
                try {
                    val variation = Variation(i)
                    fuzzee(variation)
                } catch (t: Throwable) {
                    throw IllegalStateException(String.format("Exception using variation seed %d of %d", i, times), t)
                }

            }
        }
    }
}